internal/db/rules.go

```diff
@@ -1,45 +1,175 @@
+	stmt, err := c.Conn.Prepare("INSERT INTO bounce_rule(response_code,enhanced_code,regex,priority,description,bounce_action) VALUES(?,?,?,?,?,?)")
+	
+	if err != nil {
+		log.Println(err)
+		return nil, err;
+	}
+
+
+	_, err = stmt.Exec(rule.ResponseCode, rule.EnhancedCode, rule.Regex, rule.Priority, rule.Description, rule.BounceAction)
+	
+	if err != nil {
+		log.Println(err)
+		return nil, err;
+	}
 
-	return rulesObject, nil
+	data, err := json.Marshal(rule)
+	
+	if err != nil {
+		log.Println(err)
+		return nil, err;
+	}
+	return data, nil
 }
 
-func checkErr(err error) {
+func (c *Client) GetRuleDB(id string) (*models.BounceRule, error) {
+	var bounce_rule *models.BounceRule
+	rows, err := c.Conn.Query("SELECT * FROM bounce_rule WHERE id=" + id)
+	for rows.Next() {
+		var br models.BounceRule
+		err = rows.Scan(&br.ID, &br.ResponseCode, &br.EnhancedCode, &br.Regex, &br.Priority, &br.Description, &br.BounceAction)
+		if err != nil {
+			log.Println(err)
+			return nil, err;
+		}
+		bounce_rule = &br
+		return bounce_rule, nil
+	}
+	return nil, err
+}
+
+func (c *Client) DeleteRuleDB(id int) (error){
+	query := fmt.Sprintf("%s%d", "DELETE FROM bounce_rule WHERE id=", id)
+	_, err := c.Conn.Query(query)
+
+	if err != nil {
+		log.Println(err)
+		return err;
+	}
+	return nil;
+}
+
+func (c *Client) UpdateRuleDB(ruleDifferences map[string]interface{}, prevRule *models.BounceRule) error {
+	queryString := createUpdateQuery(ruleDifferences, prevRule)
+	fmt.Println(queryString)
+	_, err :=  c.Conn.Query(queryString)
 	if err != nil {
-		fmt.Println(err)
+		fmt.Println("fails here")
+
+		log.Println(err)
+		return err;
+	}
+	return nil;
+}
+
+// getRuleDifferences this and the next functions don't require the the DB client to function do they need
+// need the (c *Client)?
+func GetRuleDifferences(prevRule *models.BounceRule, newRule *models.BounceRule) (map[string]interface{}) {
+	ruleChange := make(map[string]interface{})
+
+	if (prevRule.ID != newRule.ID){
+		ruleChange["id"] = newRule.ID
 	}
+	if (prevRule.ResponseCode != newRule.ResponseCode){
+		ruleChange["response_code"] = newRule.ResponseCode
+	}
+	if (prevRule.EnhancedCode != newRule.EnhancedCode){
+		ruleChange["enhanced_code"] = newRule.EnhancedCode
+	}
+	if (prevRule.Regex != newRule.Regex){
+		ruleChange["regex"] = newRule.Regex
+	}
+	if (prevRule.Priority != newRule.Priority){
+		ruleChange["priority"] = newRule.Priority
+	}
+	if (prevRule.Description != newRule.Description){
+		ruleChange["description"] = newRule.Description
+	}
+	if (prevRule.BounceAction != newRule.BounceAction){
+		ruleChange["bounce_action"] = newRule.BounceAction
+	}
+	return ruleChange
 }
+
+func  createUpdateQuery(ruleDifferences map[string]interface{}, prevRule *models.BounceRule) string {
+	var queryString string = "UPDATE bounce_rule SET "
+	innerCount := 1
+	for k, v := range ruleDifferences {
+
+		queryString += k + "="
+		switch v.(type) {
```

Comment:
We should avoid type inferences and instead be explicit. This function is doing a lot. I think we should simplify it. To start can you explain what is going on, and maybe give examples of input and expected output.

Reply:
GetRuleDifferences() above it takes two BounceRule structs and checks to see which items are different then returns a map of the items that are different.

CreateUpdateQuery takes that map created above and creates the update query string only updating the things that have changed. The type interface building the string

Reply:
For example if a map was `{bounce_rule: "this changed",  Priority: 123}`

this function would return 
`' UPDATE bounce_rule SET bounce_rule="this changed"  Priority: "123" where id=123'`

Reply:
This was the crud rules kenny wrote not sure why he did this way. That's the way I would have dont it would it be better to just scrap his ways and change it?